### PR TITLE
exocom - rewrite in go: broadcasting and translating messages

### PR DESCRIPTION
**Pull request by mcharawi**
@charlierudolph 
@cheahuychou 
@hugobho 
@kevgo 

Rewrite of exocom in Go.

### CODE REVIEW

File: go/structs/message.go [hunk 1]
```diff
@@ -6,11 +6,13 @@ import "time"
 type Message struct {
 	ID           string         `json:"id"`
 	Name         string         `json:"name"`
+	OriginalName string         `json:"originalName"`
```
Comment (charlierudolph) [inline go/structs/message.go]: I don't think we should have the `originalName` as part of the message struct. I'd like it to only define the fields that are sent over the websockets.

File: go/exocom/src/exocom/exocom.go [hunk 2]
```diff
@@ -106,3 +125,71 @@ func parseRegisterMessagePayload(message structs.Message) (string, error) {
...
 	return "", fmt.Errorf("Invalid register message payload: %v", message.Payload)
 }
+
+func (e *ExoCom) send(message structs.Message) error {
+	publicMessageName := messageTranslator.GetPublicMessageName(&messageTranslator.GetPublicMessageNameOptions{
+		Namespace:           e.clientRegistry.Clients[message.Sender].InternalNamespace,
+		ClientName:          message.Sender,
+		InternalMessageName: message.Name,
+	})
+	message.OriginalName = message.Name
```
Comment (charlierudolph) [inline go/exocom/src/exocom/exocom.go]: I think we can get by without setting the OriginalName as a property on the message and just passing it a variable to `sendToServices`

File: go/exocom/main_test.go [hunk 3]
```diff
@@ -153,10 +156,11 @@ func FeatureContext(s *godog.Suite) {
...
 		}
-		cmdStderr, err = cmd.StderrPipe()
-		if err != nil {
-			return err
-		}
+		cmd.Stderr = os.Stderr
+		//cmdStderr, err = cmd.StderrPipe()
+		//if err != nil {
+		//	return err
+		//}
```
Comment (charlierudolph) [inline go/exocom/main_test.go]: Please revert these changes

File: go/exocom/main_test.go [hunk 4]
```diff
@@ -196,6 +215,40 @@ func FeatureContext(s *godog.Suite) {
 		}
 		return nil
 	})
+
+	s.Step(`^ExoCom broadcasts the message "([^"]*)" to the "([^"]*)" service$`, func(messageName, serviceName string) error {
+		//TODO: chekc that the received message matches the given message
```
Comment (charlierudolph) [inline go/exocom/main_test.go]: Is there anything todo here?

File: go/exocom/main_test.go [hunk 5]
```diff
@@ -196,6 +215,40 @@ func FeatureContext(s *godog.Suite) {
...
+		_, err := services[serviceName].WaitForMessageWithName(messageName)
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (charlierudolph) [inline go/exocom/main_test.go]:
```go
if err != nil {
  return err
}
return nil

// can also be simplified to

return err

```

File: go/exocom/main_test.go [hunk 6]
```diff
@@ -28,6 +29,8 @@ func FeatureContext(s *godog.Suite) {
 	var shouldCmdBeKilled bool
 	var server *http.Server
 	var exocomPort int
+	var lastSentMessage string
+	var lastSentMessageName string
```
Comment (charlierudolph) [inline go/exocom/main_test.go]: What are these variable used for? They appear to be assigned but are otherwise unused

File: go/exocom/main_test.go [hunk 7]
```diff
@@ -196,6 +215,40 @@ func FeatureContext(s *godog.Suite) {
...
+	})
+
+	s.Step(`^ExoCom broadcasts the reply "([^"]*)" to the "([^"]*)" service$`, func(messageName, serviceName string) error {
+		//TODO: chekc that the received message matches the given message
+		_, err := services[serviceName].WaitForMessageWithName(messageName)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
```
Comment (charlierudolph) [inline go/exocom/main_test.go]: This step definition is very similar to `ExoCom broadcasts the message ...`, since we aren't doing anything different I think we update the gherkin to just use one.

File: go/exocom/main_test.go [hunk 8]
```diff
@@ -196,6 +215,40 @@ func FeatureContext(s *godog.Suite) {
...
+
+	s.Step(`^ExoCom outputs "([^"]*)"$`, func(errorText string) error {
+		return testHelpers.WaitForText(cmdStdout, errorText)
+	})
```
Comment (charlierudolph) [inline go/exocom/main_test.go]: Lets merge this and the `ExoCom signals ...` step definition by only using one in the gherkin

**Comment by mcharawi** [comment]
comments addressed @charlierudolph

File: go/exocom/src/exocom/exocom.go [hunk 9]
```diff
@@ -106,3 +125,71 @@ func parseRegisterMessagePayload(message structs.Message) (string, error) {
...
 }
+
+func (e *ExoCom) send(message structs.Message) error {
+	publicMessageName := messageTranslator.GetPublicMessageName(&messageTranslator.GetPublicMessageNameOptions{
+		Namespace:           e.clientRegistry.Clients[message.Sender].InternalNamespace,
+		ClientName:          message.Sender,
+		InternalMessageName: message.Name,
+	})
+	originalName := message.Name
+	message.Name = publicMessageName
```
Comment (charlierudolph) [inline go/exocom/src/exocom/exocom.go]: I think we can avoid overriding this as well. I'd like to try and not update message. I think we can get away with using a publicMessageName variable that gets passed into sendToServices

File: go/exocom/src/exocom/exocom.go [hunk 10]
```diff
@@ -106,3 +125,71 @@ func parseRegisterMessagePayload(message structs.Message) (string, error) {
...
+	}
+	internalMessageNameMapping, err := e.sendToServices(message, subscribers)
+	if err != nil {
+		return err
+	}
+	printLogError(e.logger.Messages(message, originalName, internalMessageNameMapping))
+	return nil
+}
+
+func (e *ExoCom) sendToServices(message structs.Message, subscribers []clientRegistry.Subscriber) (map[string]string, error) {
```
Comment (kevgo) [inline go/exocom/src/exocom/exocom.go]: What sort of map does this method return? Ideally this is documented via a variable name for the result map.

File: go/exocom/test_helpers/mock_service.go [hunk 11]
```diff
@@ -43,6 +47,15 @@ func (m *MockService) Connect() error {
 	return m.socket.WriteMessage(websocket.TextMessage, serializedBytes)
 }
 
+//Send llows the mock service to send a websocket message to exocom
```
Comment (kevgo) [inline go/exocom/test_helpers/mock_service.go]: typo in `llows`

File: go/exocom/features/broadcasting-messages.feature [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  - ExoCom sends the message to all subscribed services
+
+
+  Background:
+    Given an ExoCom instance configured with the routes:
+    """
+    [
+      {
+        "role": "web",
+        "receives": ["users.created"],
```
Comment (mcharawi) [inline go/exocom/features/broadcasting-messages.feature]: we will address this when we change how message translation works.
